src/randomness/prng.rs

```diff
@@ -0,0 +1,297 @@
+    }
+
+    pub fn reseed(&mut self, bytes: &[u8]) {
+        self.hash_chain.reseed(bytes);
+    }
+
+    pub fn random_bytes(&mut self, random_bytes_out: &mut [u8]) {
+        self.hash_chain.random_bytes(random_bytes_out);
+    }
+
+    // template <typename OtherHashT>
+    // OtherHashT RandomHash() {
+    //   return OtherHashT::InitDigestTo(RandomByteVector(OtherHashT::kDigestNumBytes));
+    // }
+
+    // pub fn random_other_hash
+
+    pub fn mix_seed_with_bytes(&mut self, raw_bytes: &[u8]) {
+        let seed_increment: u64 = 1;
+
+        self.hash_chain
+            .mix_seed_with_bytes(raw_bytes, seed_increment);
+    }
+
+    pub fn prng_state(&self) -> Vec<u8> {
+        self.hash_chain.get_hash_chain_state().to_vec()
+    }
+
+    pub fn hash_name() -> &'static str {
+        "Keccak256"
+    }
+
+    pub fn random_bytes_vec(&mut self, n_elements: usize) -> Vec<u8> {
+        let mut return_vec = vec![0u8; n_elements];
+        self.random_bytes(&mut return_vec);
+        return_vec
+    }
+
+    // TODO : implement numeric generic version e.g u8, u16, i32
+    pub fn uniform_int(&mut self, min: u64, max: u64) -> u64 {
+        assert!(min <= max, "Invalid interval");
+        let mut buf = [0u8; 8];
+        self.random_bytes(&mut buf);
+        let random_value = u64::from_le_bytes(buf);
+
+        if min == 0 && max == u64::MAX {
+            random_value
+        } else {
+            min + (random_value % (max - min + 1))
+        }
+    }
+
+    pub fn uniform_int_vec(&mut self, min: u64, max: u64, n_elements: usize) -> Vec<u64> {
+        assert!(min <= max, "Invalid interval");
+        let mut return_vec = Vec::with_capacity(n_elements);
+        for _ in 0..n_elements {
+            return_vec.push(self.uniform_int(min, max));
+        }
+        return_vec
+    }
+
+    pub fn uniform_distinct_int_vec(&mut self, min: u64, max: u64, n_elements: usize) -> Vec<u64> {
+        assert!(min <= max, "Invalid interval");
+        let n_elements_max = if max == min {
+            0
+        } else {
+            ((max - min - 1) / 2 + 1) as usize
+        };
+        assert!(
+            n_elements <= n_elements_max,
+            "Number of elements must be less than or equal to half the number of elements in the interval"
+        );
+        let mut return_vec = Vec::with_capacity(n_elements);
+        let mut current_set = HashSet::new();
+        while current_set.len() < n_elements {
+            let value = self.uniform_int(min, max);
+            if current_set.insert(value) {
+                return_vec.push(value);
+            }
+        }
+        return_vec
+    }
+
+    // XXX : dumb implementation. should reduce calls to uniform_int_vec
+    pub fn uniform_bool_vec(&mut self, n_elements: usize) -> Vec<bool> {
+        let bits = self.uniform_int_vec(0, 1, n_elements);
+        bits.into_iter().map(|bit| bit != 0).collect()
+    }
+
+    pub fn uniform_bigint<B: BigInteger>(&mut self, min: B, max: B) -> B {
```

Comment:
This corresponds to https://github.com/starkware-libs/stone-prover/blob/main/src/starkware/algebra/fields/long_field_element.cc#L61 right?

Reply:
This corresponds to https://github.com/starkware-libs/stone-prover/blob/main/src/starkware/randomness/prng.h#L147